Use borrow trait to add items into sketch

SuperFluffy opened the pull request:
Using the borrow trait when inserting a value into the sketch, i.e. `value: &Q` with `H: Borrow<Q>` rather than `value: &H`, makes the sketch types much more flexible. For example, you don't run into the issue of having a type that contains a lifetime in the sketch: `struct HyperLogLog<H>` with `H = MyType<'a>` causes issues when `MyType<'a>` is a temporary value that references a type that's being destroyed at the end of some scope.

I ran into the issue where I don't insert “owned” items into the sketch, but something like `MyType<'a>` instead, and I quickly ran into trouble with the thing that `MyType` was referencing being destroyed. That's because `HyperLogLog<MyType<'a>>` extended the lifetime of `'a` to that of `HyperLogLog` which wasn't workable.

This basically follows the same design as `HashMap`.

SuperFluffy committed:
Use borrow trait to add items into sketch

Using the borrow trait when inserting a value into the sketch, i.e.
`value: &Q` with `H: Borrow<Q>` rather than `value: &H`, makes the
sketch types much more flexible. For example, you don't run into the
issue of having a type that contains a lifetime in the sketch:
`struct HyperLogLog<H>` with `H = MyType<'a>` causes issues when
`MyType<'a>` is a temporary value that references a type that's being
destroyed at the end of some scope.

tabac commented:
Thanks for the PR! Could you give a small example with the case you mention? 

My only concern is that we lose type inference and I am not sure if we can avoid it.

SuperFluffy commented:
The clearest (and common) example is probably tokenizing a String and hashing it into the filter. Let's say you have some `s: String` and you apply `s.split_whitespace()` to it. This gives you an iterator that yields `&'a str` with `'a` the lifetime of `&'a String` as dereferenced via `Deref<Target=str>`.

If you have set `HyperLogLog<String>`, you will not be able to `HyperLogLog::add(token: &str)`, because `String != str`. At the same time, if you define `HyperLogLog<&'a str>` you run into lifetime issues should the filter ever want to live longer than the `String` that the tokens were derived from. This is also a very reasonable assumption; imagine passing a `&mut HyperLogLog<&'a str>` into a function that reads a file into a string, then chops it up. This ties the lifetime `'a` to the lifetime of the string, which gets dropped at the end of the function (well you couldn't actually put `'a` on anything, but for illustrative purposes I think this example is fine).


When using the library I actually ran into an additional issue that caused me to forego the parameter altogether. I want to add `&'b [&'a str] where 'a: 'b` into the sketch. The “fully owned” version of this would be `Vec<String>`, but the `Borrow` trait only allows me to go `Vec<String> -> &[String]`. I could now add an extra bound `T: Borrow<Q>, Q: AsRef<P>`, but that's turning into a whackamole.

Since the sketch doesn't actually store anything, I have decided that it would be reasonable to have the user ensure that only correct items get put into it. So I have this instead:

```rust
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct HyperLogLogPF<B>
where
    B: BuildHasher,
{
    builder:   B,
    count:     usize,
    precision: u8,
    registers: Registers,
}

impl<B> HyperLogLog for HyperLogLogPF<B>
where
    B: BuildHasher,
{
    /// Adds a new value to the multiset.
    fn add<H: Hash>(&mut self, value: &H) {
        todo!()
    }
}
```


Would you be open to a PR that drops the param altogether?

tabac commented:
Thanks for the explanation. 

I am not sure about dropping the type parameter. The restriction on the value type doesn't look unreasonable to me; I don't see any use cases we might want to insert values of different type to the sketch and it can eliminate some programming errors.

However, modifying the API to accept types implementing `Borrow` can be useful; I prepared a PR (#12) about this to deprecate the current `add` function and introduce an `insert` with the new signature.

SuperFluffy commented:
I wrote a `CountingBloomFilter<T>` and also introduced a type parameter enforced by `PhantomData<T>` with the same thought process as you had, i.e. `CountingBloomFilter::insert(&mut self, value: &T)`. This evolved into `value: &U where T: Borrow<Q>`, and eventually I dropped `T` altogether because I couldn't insert `&[&str]` into it. Unfortunately you have no way of expressing the relation `Vec<String> -> &[&str]` in Rust (outside of some. And you also don't have a way to express “hash equivalence” (i.e. think about the (partial) relations`PartialEq` or `Eq` but not on two types `T` and `U` but on their hashes `hash(T)` and `hash(U)`).

I will have to maintain a (private) fork of `hyperloglog.rs` then because I need to be able to insert `&[&str]` (well, actually `ArrayVec<[&str; 32]>` to be precise). :-/

tabac commented:
You're probably right, I guess this will come up again sooner or later. 

I merged https://github.com/tabac/hyperloglog.rs/pull/12 which doesn't change the original API but adds an `insert_any()` method to each sketch which ignores its type. It should work for your case.
